File: .claude/hooks/stop-dead-link-checker.mjs
#!/usr/bin/env node
// @hookstack stop-dead-link-checker
// Vérifie les liens relatifs cassés dans tous les fichiers Markdown du repo (Stop).
// Scan complet — couvre la dette existante, pas seulement les fichiers de la session.
// Purement Node.js (fs + path), sans réseau ni dépendance externe.
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const SKIP_DIRS = new Set([
	"node_modules",
	".git",
	"dist",
	"build",
	".next",
	"out",
	".claude",
]);
// Capture [text](href) — exclut les images ![alt](src) incluses dans la même syntaxe
const LINK_RE = /(?<!!)\[([^\]]*)\]\(([^)]+)\)/g;

function isRelative(href) {
	return (
		!href.startsWith("http") &&
		!href.startsWith("#") &&
		!href.startsWith("mailto:")
	);
}

function stripAnchor(href) {
	return href.split("#")[0].trim();
}

// Ignore les liens d'exemple à l'intérieur des blocs de code (```...```), ce ne sont
// pas de vrais liens à résoudre (ex. templates documentant la syntaxe).
function stripFences(text) {
	let inFence = false;
	return text
		.split("\n")
		.filter((line) => {
			if (/^(```|~~~)/.test(line.trim())) {
				inFence = !inFence;
				return false;
			}
			return !inFence;
		})
		.join("\n");
}

// Liens commençant par "/" : convention OKF v0.1, relatifs à la racine du bundle okf/
// (même résolution que resolveTarget() dans scripts/okf.mjs), pas au système de fichiers.
function resolveAbs(file, target, projectDir) {
	if (target.startsWith("/")) return join(projectDir, "okf", target.slice(1));
	return resolve(dirname(file), target);
}

function walkMd(dir, { readdir = readdirSync, exists = existsSync } = {}) {
	if (!exists(dir)) return [];
	const results = [];
	for (const entry of readdir(dir, { withFileTypes: true })) {
		if (SKIP_DIRS.has(entry.name)) continue;
		const full = join(dir, entry.name);
		if (entry.isDirectory()) results.push(...walkMd(full, { readdir, exists }));
		else if (/\.mdx?$/.test(entry.name)) results.push(full);
	}
	return results;
}

export function run(
	_input,
	{
		projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
		readFile = readFileSync,
		exists = existsSync,
		readdir = readdirSync,
	} = {},
) {
	const mdFiles = walkMd(projectDir, { readdir, exists });
	if (!mdFiles.length) return null;

	const broken = [];
	for (const file of mdFiles) {
		let content;
		try {
			content = readFile(file, "utf8");
		} catch {
			continue;
		}

		for (const [, , href] of stripFences(content).matchAll(LINK_RE)) {
			if (!isRelative(href)) continue;
			const target = stripAnchor(href);
			if (!target) continue; // lien ancre pure (#section)
			const abs = resolveAbs(file, target, projectDir);
			if (!exists(abs)) {
				broken.push(`${file.replace(`${projectDir}/`, "")}  →  ${href}`);
			}
		}
	}

	if (!broken.length) return null;

	return {
		message:
			`[dead-link-checker] ${broken.length} broken relative link(s) across docs:\n` +
			broken.map((b) => `  - ${b}`).join("\n") +
			"\n",
	};
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
	const input = JSON.parse(readFileSync(0, "utf8"));
	const result = run(input);
	if (result) process.stderr.write(JSON.stringify(result));
}

File: .claude/hooks/stop-dead-link-checker.py
#!/usr/bin/env python3
# @hookstack stop-dead-link-checker
"""Checks broken relative links across all Markdown files in the repo (Stop)."""
import json
import os
import re
import sys

SKIP_DIRS = {
    "node_modules",
    ".git",
    "dist",
    "build",
    ".next",
    "out",
    ".claude",
}
# Capture [text](href) — exclut les images ![alt](src) incluses dans la même syntaxe
LINK_RE = re.compile(r"(?<!!)\[([^\]]*)\]\(([^)]+)\)")


def _read_file(path):
    with open(path, "r", encoding="utf8") as f:
        return f.read()


def _exists(path):
    return os.path.exists(path)


def _readdir(path):
    with os.scandir(path) as it:
        return list(it)


def is_relative(href):
    return not href.startswith("http") and not href.startswith("#") and not href.startswith("mailto:")


def strip_anchor(href):
    return href.split("#")[0].strip()


# Ignore les liens d'exemple à l'intérieur des blocs de code (```...```)
def strip_fences(text):
    in_fence = False
    lines = []
    for line in text.split("\n"):
        if re.match(r"^(```|~~~)", line.strip()):
            in_fence = not in_fence
            continue
        if not in_fence:
            lines.append(line)
    return "\n".join(lines)


# Liens commençant par "/" : convention OKF v0.1, relatifs à la racine du bundle okf/
def resolve_abs(file, target, project_dir):
    if target.startswith("/"):
        return os.path.join(project_dir, "okf", target[1:])
    return os.path.normpath(os.path.join(os.path.dirname(file), target))


def walk_md(directory, *, readdir, exists):
    if not exists(directory):
        return []
    results = []
    try:
        entries = readdir(directory)
    except Exception:
        return results
    for entry in entries:
        if entry.name in SKIP_DIRS:
            continue
        full = os.path.join(directory, entry.name)
        if entry.is_dir():
            results.extend(walk_md(full, readdir=readdir, exists=exists))
        elif re.search(r"\.mdx?$", entry.name):
            results.append(full)
    return results


def run(_input=None, *, project_dir=None, read_file=None, exists=None, readdir=None):
    if read_file is None:
        read_file = _read_file
    if exists is None:
        exists = _exists
    if readdir is None:
        readdir = _readdir
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR", os.getcwd())

    md_files = walk_md(project_dir, readdir=readdir, exists=exists)
    if not md_files:
        return None

    broken = []
    for file in md_files:
        try:
            content = read_file(file)
        except Exception:
            continue

        for m in LINK_RE.finditer(strip_fences(content)):
            href = m.group(2)
            if not is_relative(href):
                continue
            target = strip_anchor(href)
            if not target:
                continue  # lien ancre pure (#section)
            abs_path = resolve_abs(file, target, project_dir)
            if not exists(abs_path):
                broken.append(f"{file.replace(project_dir + '/', '')}  →  {href}")

    if not broken:
        return None

    return {
        "message": (
            f"[dead-link-checker] {len(broken)} broken relative link(s) across docs:\n"
            + "\n".join(f"  - {b}" for b in broken)
            + "\n"
        )
    }


if __name__ == "__main__":
    data = json.load(sys.stdin) if not sys.stdin.isatty() else None
    result = run(data)
    if result:
        sys.stderr.write(json.dumps(result))
